hoomd/mpcd/RejectionVirtualParticleFiller.h

```diff
@@ -0,0 +1,175 @@
+        std::shared_ptr<const Geometry> getGeometry() const
+            {
+            return m_geom;
+            }
+
+        //! Set the streaming geometry
+        void setGeometry(std::shared_ptr<const Geometry> geom)
+            {
+            m_geom = geom;
+            }
+
+        //! Fill the particles outside the confinement
+        void fill(unsigned int timestep);
+
+    protected:
+        std::shared_ptr<const Geometry> m_geom;
+        GPUArray<Scalar4> m_tmp_pos;
+        GPUArray<Scalar4> m_tmp_velTag;
+    };
+
+
+template<class Geometry>
+void RejectionVirtualParticleFiller<Geometry>::fill(unsigned int timestep)
+    {
+    // Number of particles that we need to draw (constant)
+    const BoxDim& box = m_pdata->getBox();
+    const Scalar3 lo = box.getLo();
+    const Scalar3 hi = box.getHi();
+    const unsigned int NVirtMax = round(m_density*box.getVolume());
+
+    // Step 1: Create temporary GPUArrays to draw Particles locally using the worst case estimate for number
+    //         number of particles.
+    if (NVirtMax > m_tmp_pos.getNumElements())
+        {
+        GPUArray<Scalar4> tmp_pos(NVirtMax, m_exec_conf);
+        GPUArray<Scalar4> tmp_velTag(NVirtMax, m_exec_conf);
+        m_tmp_pos.swap(tmp_pos);
+        m_tmp_velTag.swap(tmp_velTag);
+        }
+    ArrayHandle<Scalar4> h_tmp_pos(m_tmp_pos, access_location::host, access_mode::overwrite);
+    ArrayHandle<Scalar4> h_tmp_velTag(m_tmp_velTag, access_location::host, access_mode::overwrite);
+
+    // Step 2: Draw the particles and assign velocities simultaneously by using temporary memory
+    unsigned int pidx = 0;
+    unsigned int tag = computeFirstTag(&pidx);
+
+    const Scalar vel_factor = fast::sqrt(m_T->getValue(timestep) / m_mpcd_pdata->getMass());
+
+    for (unsigned int i=0; i < NVirtMax; ++i)
+        {
+        hoomd::RandomGenerator rng(hoomd::RNGIdentifier::RejectionFiller, m_seed, timestep, tag, m_filler_id);
+
+        Scalar3 particle = make_scalar3(hoomd::UniformDistribution<Scalar>(lo.x, hi.x)(rng),
+                                       hoomd::UniformDistribution<Scalar>(lo.y, hi.y)(rng),
+                                       hoomd::UniformDistribution<Scalar>(lo.z, hi.z)(rng));
+
+        if (m_geom->isOutside(particle))
+            {
+            h_tmp_pos.data[pidx] = make_scalar4(particle.x,
+                                                particle.y,
+                                                particle.z,
+                                                __int_as_scalar(m_type));
+
+            hoomd::NormalDistribution<Scalar> gen(vel_factor, 0.0);
+            Scalar3 vel;
+            gen(vel.x, vel.y, rng);
+            vel.z = gen(rng);
+            h_tmp_velTag.data[pidx++] = make_scalar4(vel.x,
+                                                     vel.y,
+                                                     vel.z,
+                                                     __int_as_scalar(tag));
+
+            tag++;
```

Comment:
We'll compact the tags later, can just set this to no-cell now
```suggestion
                                                     __int_as_scalar(mpcd::detail::NO_CELL));
```